### PR TITLE
Add `goto-bmc`, a new tool to run only the verifier

### PR DESCRIPTION
**Pull request by NlightNFotis**
This adds a new tool, `goto-bmc`, which you can think of as `CBMC`-lite: its purpose is to run only the backend, on a binary that is produced to be [`symex-ready`](https://diffblue.github.io/cbmc/adr/symex-ready-goto.html), i.e. have been lowered to a form that is accepted by symex without the need for any more transformations.

The aim of the addition of this tool is to contribute towards the efforts to make `CBMC` and all of the assorted tools more modularised, by providing access to the backend of CBMC without needing to bundle frontends, transformation passes, etc, so that usage or integration of the tool into other software or (CI/dev) pipelines becomes easier.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/libcprover-cpp/api.h [hunk 1]
```diff
@@ -98,7 +98,8 @@ struct api_sessiont
 
   /// Implement necessary transformations to reduce model to core-GOTO, before
   /// being fed to symex.
-  /// @return The function return `true` if it failed because CBMC produced an error.
+  /// @return The function return `true` if it failed because CBMC produced
```
Comment (TGWDB) [inline src/libcprover-cpp/api.h]: return -> returns

File: src/libcprover-cpp/verification_result.cpp [hunk 2]
```diff
@@ -149,3 +150,21 @@ verification_resultt::get_property_status(const std::string &property_id) const
   }
   UNREACHABLE;
 }
+
+// FOTIS' note: Modelled after `result_to_exit_code` in
+// `src/goto-checker/properties.cpp`.
```
Comment (TGWDB) [inline src/libcprover-cpp/verification_result.cpp:155]: Should this be refactored to use the same code then?

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+#include "api.h"
+
+#include <iostream>
+#include <map>
+
+core_cbmc_only_parse_optionst::core_cbmc_only_parse_optionst(
+  int argc,
+  const char **argv)
+  : parse_options_baset(
+      CORE_CBMC_ONLY_OPTIONS,
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: symex-ready

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      argc,
+      argv,
+      std::string("CORE_CBMC_ONLY") + CBMC_VERSION)
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: symex-ready

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+void print_messages_to_stdout(
+  const api_messaget &message,
+  api_call_back_contextt)
+{
+  std::cout << api_message_get_string(message) << std::endl;
+}
+
+int core_cbmc_only_parse_optionst::doit()
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: symex-ready

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  std::string filename = cmdline.args[0];
+
+  if(cmdline.isset("validate-goto-model"))
+  {
+    api_options.validate_goto_model(true);
+  }
+
+  // With options set, we can now initiate the API and perform analysis.
+  api_sessiont api{api_options};
+  std::cout << "CoreBMC version ";
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: goto-bmc?

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  // Finally, we read the goto-binary the user supplied...
+  api.read_goto_binary(filename);
+
+  // ... and run analysis on it.
+  auto result = api.run_verifier();
+
+  return verifier_result_to_exit_code(result->final_result());
+}
+
+void core_cbmc_only_parse_optionst::help()
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: symex-ready

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            << align_center_with_border("Copyright (C) 2023") << '\n'
+            << align_center_with_border("Diffblue Ltd.") << '\n' // NOLINT(*)
+            <<
+    "\n"
+    "Usage:                       Purpose:\n"
+    "\n"
+    "goto-bmc [-?] [-h] [--help]      show help\n"
+    "goto-bmc --version               show version and exit\n"
+    // NOLINTNEXTLINE(*)
+    "goto-bmc [options] file.c ...    perform bounded model checking on core goto-binary\n";
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: symex-ready

File: src/goto-bmc/goto_bmc_parse_options.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+#include <memory>
+#include <string>
+#include <vector>
+
+api_optionst parse_command_line_options(const std::vector<std::string> &);
+
+void help();
+
+// clang-format off
+#define CORE_CBMC_ONLY_OPTIONS \
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.h]: symex-ready

File: src/goto-bmc/goto_bmc_parse_options.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  "(version)"
+// clang-format on
+
+class core_cbmc_only_parse_optionst : public parse_options_baset
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.h]: symex-ready

File: src/goto-bmc/goto_bmc_parse_options.h [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+{
+public:
+  core_cbmc_only_parse_optionst(int argc, const char **argv);
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.h]: symex-ready

File: src/goto-bmc/goto_bmc_parse_options.h [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+  virtual void help();
+
+  virtual int doit();
+};
+
+#endif // CPROVER_CORE_BMC_PARSE_OPTIONS_H
```
Comment (TGWDB) [inline src/goto-bmc/goto_bmc_parse_options.h]: symex-ready

File: src/libcprover-cpp/api.h [hunk 13]
```diff
@@ -79,11 +79,28 @@ struct api_sessiont
...
+  /// A simple API version information function.
   std::unique_ptr<std::string> get_api_version() const;
 
+  /// Process the model by running symex and the decision procedure.
+  /// @return a `unique_ptr` to the `verification_resultt` summary.
+  std::unique_ptr<verification_resultt> run_verifier() const;
+
+  /// Read a goto-binary from a given filename.
+  /// @warning Will error out if it reads a source file.
+  void read_goto_binary(std::string &file) const;
```
Comment (TGWDB) [inline src/libcprover-cpp/api.h:94]: This should check that the goto binary is in symex-ready form. Of course the check is not implemented yet, but maybe this should be a comment or stub in the implementation?

File: src/libcprover-cpp/api.h [hunk 14]
```diff
@@ -79,11 +79,28 @@ struct api_sessiont
...
+
+  /// True if file is goto-binary.
+  bool is_goto_binary(std::string &file) const;
```
Comment (TGWDB) [inline src/libcprover-cpp/api.h:97]: Maybe here,  -> `is_symex_ready_goto_binary`?

File: src/libcprover-cpp/api.h [hunk 15]
```diff
@@ -79,11 +79,28 @@ struct api_sessiont
...
+
 private:
   std::unique_ptr<api_session_implementationt> implementation;
+
+  /// Implement necessary transformations to reduce model to core-GOTO, before
```
Comment (TGWDB) [inline src/libcprover-cpp/api.h]: symex-ready

Comment (NlightNFotis) [reply src/libcprover-cpp/api.h]: Done.

File: src/goto-bmc/goto_bmc_parse_options.h (hunk 12 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.h:34]: This is refering to the tool name (`core-bmc`).

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 8 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: Done.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 3 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: These shouldn't be called `symex-ready` - they refer to the tool not the goto-binaries.

But they will need to be made consistent to the actual tool name, which we settled on `goto-bmc` (this is the name of the tool built). The references are artefacts from a previous change where we called the tool `corebmc` but settled on `goto-bmc` instead.

File: src/libcprover-cpp/api.h (hunk 14 above)
Comment (NlightNFotis) [reply src/libcprover-cpp/api.h:97]: We don't want to do this, because we delegate to `is_goto_binary` from `src/goto-programs/read_goto_binary.cpp` which doesn't check the "level" of the goto-binary (whether transformations have been performed and which ones), so we can't really promise to give information we don't have.

File: src/libcprover-cpp/api.h (hunk 13 above)
Comment (NlightNFotis) [reply src/libcprover-cpp/api.h:94]: We track this functionality, by having a negative test under `regression/goto-bmc/goto-bmc-non-symex-ready-goto` whose job is to fail when fed a non-symex-ready goto.

For now it fails with an invariant violation, so the error message leaves a lot to be desired, but when we fix the recognition of the symex-ready-goto binaries, this test is going to fail (because the invariant violation pattern is not going to be present in the output any more) and this will serve as our reminder to go back and fix that.

File: src/libcprover-cpp/verification_result.cpp (hunk 2 above)
Comment (NlightNFotis) [reply src/libcprover-cpp/verification_result.cpp:155]: We can't do that because the functions use different types as inputs.

We have two types that are equivalent, but one of them is internal to CBMC (`goto-checker`, really), and thus, we can't expose it to the C++ API (because of the internal headers issue), and the other is the C++ API equivalent, which is present in the headers for the API.

We have in the API (under `verification_result.cpp`) a function which does the translation from `result_internal -> result_external`, but we needed to have a function from `result_external -> BMC_EXIT_CODE`. This is the function that was modelled under the equivalent one from `result_internal -> BMC_EXIT_CODE`.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 6 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: Done.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 3 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: Done.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 4 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: Renamed to `goto-bmc` - Done.

**Review by TGWDB — APPROVED** [review]

File: src/libcprover-cpp/api.cpp [hunk 16]
```diff
@@ -137,6 +137,18 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
...
 
+  bool empty_result = preprocess_model();
+  if (empty_result) {
+    return {};
+  }
+
+  return run_verifier();
+}
+
+/// <FILL ME>
```
Comment (peterschrammel) [inline src/libcprover-cpp/api.cpp]: ❓

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+goto_bmc_parse_optionst::goto_bmc_parse_optionst(int argc, const char **argv)
+  : parse_options_baset(
+      GOTO_BMC_OPTIONS,
+      argc,
+      argv,
+      std::string("goto-bmc") + CBMC_VERSION)
+{
+}
+
+void print_messages_to_stdout(
```
Comment (peterschrammel) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: ❓ What is this function for? 
We should use `log` for printing messages.

File: src/goto-bmc/module_dependencies.txt [hunk 18]
```diff
@@ -0,0 +1,3 @@
+util
+libcprover-cpp
+goto-programs
```
Comment (peterschrammel) [inline src/goto-bmc/module_dependencies.txt]: Why this dependency? Shouldn't goto-bmc only depend on libcprover-cpp?

Comment (NlightNFotis) [reply src/goto-bmc/module_dependencies.txt]: We need this because of the check inside `src/goto-bmc/goto_bmc_parse_options.cpp:52` which has a call to `is_goto_binary`, defined under `src/goto-programs/read_goto_binary.cpp`

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 17 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: The way the API works for now is it needs a callback to be passed to it, which it then calls to process messages from it.

This callback function needs to have a specific signature for it to be able to be registered with the API.

What we *can* do however, is change the references to `std::cout` to output things via the `log` instead.

File: src/libcprover-cpp/api.cpp (hunk 16 above)
Comment (NlightNFotis) [reply src/libcprover-cpp/api.cpp]: This was empty while we did some refactoring, acting as a reminder to fill in documentation once the code stabilised - it's filled in the next commit.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 17 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: Done in 755f0f555721ea372b93f5a1bccef2b6513969c8

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+  const api_messaget &message,
+  api_call_back_contextt)
+{
+  // TODO: this is just a hack while a better way to connect the
+  //  parse_options_baset::log to the api callback is found.
+  //  This has to be removed before merging the PR.
+  ui_message_handlert ui_message_handler(cmdlinet{}, "goto-bmc");
+  messaget log{ui_message_handler};
```
Comment (esteffin) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: This has to be fixed properly.
However the older way to create a log by just calling `messaget log;` was not working.

**Review by esteffin — CHANGES_REQUESTED** [review]
The log-handling function has to be fixed properly

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 17 above)
Comment (thomasspriggs) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: > Done in [755f0f5](https://github.com/diffblue/cbmc/commit/755f0f555721ea372b93f5a1bccef2b6513969c8)

Is that actually a good idea? With the new API the log exists inside the API layer but is not exposed. So by creating a log in the entry point doesn't that mean two instances of it exist at once?

File: src/goto-bmc/module_dependencies.txt (hunk 18 above)
Comment (peterschrammel) [reply src/goto-bmc/module_dependencies.txt]: If you need it then the API needs to expose it, no?

Comment (NlightNFotis) [reply src/goto-bmc/module_dependencies.txt]: I'm wrong, we already expose it in the API - we will change it to make it defer to the API (we already have this capability, probably lost the idea somewhere during refactoring)

File: src/libcprover-cpp/api.h [hunk 20]
```diff
@@ -79,11 +79,17 @@ struct api_sessiont
   /// Validate the loaded goto model
   void validate_goto_model() const;
 
-  // A simple API version information function.
+  /// A simple API version information function.
   std::unique_ptr<std::string> get_api_version() const;
 
+  /// Run the verifier! <FILL ME>
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.h]: :pick: Shouldn't the documentation be included in the commit which adds the function?

File: src/libcprover-cpp/api.cpp [hunk 21]
```diff
@@ -150,11 +162,11 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
     cprover_c_library_factory);
 
   // Common removal of types and complex constructs
-  if(::process_goto_program(
-       *implementation->model, *implementation->options, log))
-  {
-    return {};
-  }
+    if(::process_goto_program(
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp]: :pick: This indenting doesn't look correct. If clang-format complains about this then the corresponding clang-format update should be in this commit and not a later one.

File: src/libcprover-cpp/api.h [hunk 22]
```diff
@@ -79,11 +79,17 @@ struct api_sessiont
   /// Validate the loaded goto model
   void validate_goto_model() const;
 
-  // A simple API version information function.
+  /// A simple API version information function.
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.h:85]: :pick: Fix of existing documentation should be in a separate commit from the refactor, if the function is not changed as part of the refactor.

File: src/libcprover-cpp/api.h [hunk 23]
```diff
@@ -79,11 +79,17 @@ struct api_sessiont
...
   std::unique_ptr<std::string> get_api_version() const;
 
+  /// Run the verifier! <FILL ME>
+  std::unique_ptr<verification_resultt> run_verifier() const;
+
 private:
   std::unique_ptr<api_session_implementationt> implementation;
+
+  /// <FILL ME>
+  bool preprocess_model() const;
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.h:106]: :bulb: If this function is truly an implementation detail, then it would make sense to me for it to be a free function which is defined the `.cpp` only.
:pick: If the plan is to eventually use this as part of the publicly facing API then it isn't following the existing code pattern, which is to use the call-back to indicate errors and to return a `void` result.

File: src/libcprover-cpp/api.cpp [hunk 24]
```diff
@@ -138,15 +139,25 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
   PRECONDITION(implementation->model);
 
   bool empty_result = preprocess_model();
-  if (empty_result) {
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp]: :pick: Why is this formatting change in this commit? Comment applies to any other comment on this commit marked with a :foot:

File: src/libcprover-cpp/api.cpp [hunk 25]
```diff
@@ -162,11 +173,11 @@ bool api_sessiont::preprocess_model() const
     cprover_c_library_factory);
 
   // Common removal of types and complex constructs
-    if(::process_goto_program(
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp]: :foot:

File: src/libcprover-cpp/api.cpp [hunk 26]
```diff
@@ -183,19 +194,21 @@ bool api_sessiont::preprocess_model() const
...
 }
 
-/// <FILL ME>
 std::unique_ptr<verification_resultt> api_sessiont::run_verifier() const
 {
   ui_message_handlert ui_message_handler(*implementation->message_handler);
   all_properties_verifier_with_trace_storaget<multi_path_symex_checkert>
     verifier(
       *implementation->options, ui_message_handler, *implementation->model);
+
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp:220]: :foot:

File: src/libcprover-cpp/api.cpp [hunk 27]
```diff
@@ -183,19 +194,21 @@ bool api_sessiont::preprocess_model() const
...
   verification_resultt result;
   auto results = verifier();
+
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp:223]: :foot:

File: src/libcprover-cpp/api.cpp [hunk 28]
```diff
@@ -183,19 +194,21 @@ bool api_sessiont::preprocess_model() const
...
   // Print messages collected to callback buffer.
   verifier.report();
   // Set results object before returning.
   result.set_result(results);
+
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp:228]: :foot:

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+  api_sessiont api{api_options};
+  std::cout << "goto-bmc version ";
+  std::cout << *api.get_api_version() << std::endl;
+
+  // The API works with a callback for querying state - we for now just print
+  // collected messages from the message buffer to stdout.
+  api.set_message_callback(print_messages_to_stdout, nullptr);
+
+  // Finally, we read the goto-binary the user supplied...
+  api.read_goto_binary(filename);
```
Comment (thomasspriggs) [inline src/goto-bmc/goto_bmc_parse_options.cpp:72]: :no_entry_sign: Missing error handling. I am thinking about handling of basic issues like the specified file not existing. We should not just violate an INVARIANT and dump a stack on bad input from the user. This could be resolved via the message_call back interface via the following steps -

1. Add a `message_levelt` to `api_messaget`.
2. Populate it in `api_message_handlert::print`, using the `level` passed in.
3. Add corresponding getter to the api, which gets the error status from an `api_messaget`.
4. Update the message call back used in this entry point such that the callback context includes an error_encountered flag, which is set to `true` when an error message is sent through the call back.
5. Add exit on error flag tripped.

File: src/goto-bmc/goto_bmc_parse_options.h [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+#include <util/parse_options.h>
+#include <util/version.h>
+
+#include <libcprover-cpp/api_options.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+api_optionst parse_command_line_options(const std::vector<std::string> &);
```
Comment (thomasspriggs) [inline src/goto-bmc/goto_bmc_parse_options.h]: :no_entry_sign: This `parse_command_line_options` function seems to be unimplemented at this point in the commit history. The same comment applies to `void help();` Therefore the forward declaration should not exist in this commit.

File: src/goto-bmc/goto_bmc_parse_options.h [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+#define GOTO_BMC_OPTIONS \
+  "(version)"
+// clang-format on
+
+class goto_bmc_parse_optionst : public parse_options_baset
+{
+public:
+  goto_bmc_parse_optionst(int argc, const char **argv);
+
+  virtual void help();
```
Comment (thomasspriggs) [inline src/goto-bmc/goto_bmc_parse_options.h]: :hammer: As we are overriding the `help` function from `parse_options_baset`, this should be specified as `void help() override;` without `virtual`. Note that this is standardised in `CODING_STANDARD.md` line 196.
This comment also applies to the `doit` function just below.

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 32]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
+  std::string filename = cmdline.args[0];
+
+  if(cmdline.isset("validate-goto-model"))
+  {
+    api_options.validate_goto_model(true);
+  }
+
+  // With options set, we can now initiate the API and perform analysis.
```
Comment (thomasspriggs) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: :no_entry_sign: This comment is misleading, because the block of code immediately below it does not perform the analysis. 
:pick: Also is a comment which says "With options set, we can now initiate the API" useful, when the line below it is `api_sessiont api{api_options};` which says initialise the API session using the options (in C++)? IMO the code should tell you what the code does. With well named code there should be no need to translate C++ to English and add the English to a comment. It just leaves scope for the English to mismatch the C++. Comments are really valuable when they explain _**why**_ the code is written the way it is, not just existing as a duplication of what the reader can see for themselves.

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 33]
```diff
@@ -0,0 +1,97 @@
...
+  api_sessiont api{api_options};
+  std::cout << "goto-bmc version ";
+  std::cout << *api.get_api_version() << std::endl;
+
+  // The API works with a callback for querying state - we for now just print
+  // collected messages from the message buffer to stdout.
```
Comment (thomasspriggs) [inline src/goto-bmc/goto_bmc_parse_options.cpp:52]: :information_source: A capture-less lambda can be converted to a function pointer (without using `std::function`). This technique could be used in order to keep the definition of the function next-to/in-line-with the setting of the message callback. This would restrict the scope of the function and might dodge such reader questions as "what is this function for?"

File: src/libcprover-cpp/verification_result.h [hunk 34]
```diff
@@ -80,4 +80,8 @@ struct verification_resultt
   std::unique_ptr<verification_result_implt> _impl;
 };
 
+// Allow translation of `verifier_resultt` into a CPROVER_EXIT_CODES (so that
```
Comment (thomasspriggs) [inline src/libcprover-cpp/verification_result.h:83]: :pick: The use of a plural `CPROVER_EXIT_CODES` suggests one result is turned into multiple codes.
:thinking: Given that the API headers may be distributed without the rest of the project source, the reader maybe unaware of the existence of `src/util/exit_codes.h`. So maybe we should just call this a Command Line Interface (CLI) return value/exit code; especially given that the result is not strongly typed.

File: src/libcprover-cpp/api.cpp [hunk 35]
```diff
@@ -137,6 +138,28 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
...
+    return {};
+  }
+
+  return run_verifier();
+}
+
+void api_sessiont::read_goto_binary(std::string &file) const
+{
+  implementation->model = util_make_unique<goto_modelt>(
+    ::read_goto_binary(file, *implementation->message_handler).value());
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp]: :no_entry_sign: `read_goto_binary` may return an empty optional. In that case this implementation will call `.value()` on it to get the non-existent value; which isn't exactly going to fail gracefully.

File: src/libcprover-cpp/api.cpp [hunk 36]
```diff
@@ -137,6 +138,28 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
...
+}
+
+bool api_sessiont::is_goto_binary(std::string &file) const
```
Comment (thomasspriggs) [inline src/libcprover-cpp/api.cpp:173]: :grey_question: Please consider whether having a separate `read_goto_binary` function and a `is_goto_binary` function is actually a good interface for a publicly facing API? This kind of implies that the file has to be opened and closed twice for the typical use case.

File: src/goto-bmc/goto_bmc_parse_options.cpp [hunk 37]
```diff
@@ -48,7 +48,7 @@ int goto_bmc_parse_optionst::doit()
   }
 
   if(
-    cmdline.args.size() != 1 ||
+    cmdline.args.size() > 1 &&
```
Comment (thomasspriggs) [inline src/goto-bmc/goto_bmc_parse_options.cpp]: This change appears to be in the wrong commit, given that the commit title refers to regression tests, but this change is a change to the implementation.

File: regression/goto-bmc/goto-bmc-non-symex-ready-goto/negative1/main.desc [hunk 38]
```diff
@@ -1,7 +1,7 @@
 CORE
 main.c
 
-^EXIT=[127|134]$
```
Comment (thomasspriggs) [inline regression/goto-bmc/goto-bmc-non-symex-ready-goto/negative1/main.desc]: :pick: The commit title refers to moving tests, but this part is changing the test.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 32 above)
Comment (NlightNFotis) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: Yeah, that has been removed, but I believe in a commit later than the one it was commented on. Agree with the comment in the review window (the original commits are there to guide our implementation during the green-field implementation, and they fell through the cracks during various refactorings/handoffs of the code).

File: src/libcprover-cpp/api.cpp (hunk 36 above)
Comment (NlightNFotis) [reply src/libcprover-cpp/api.cpp:173]: This requires some more thinking as to how it's positioned in the API, but the `is_goto_binary` opens the file and reads just 8 bytes of it (a header) to determine whether it's a goto-binary, so it doesn't result in much inefficiency (if I understand your concerns).

From a code organisation standpoint, yeah, I need to think about it a bit more.

File: src/libcprover-cpp/api.cpp (hunk 35 above)
Comment (NlightNFotis) [reply src/libcprover-cpp/api.cpp]: Done (error handling added in the first commit)

File: src/libcprover-cpp/api.cpp (hunk 36 above)
Comment (thomasspriggs) [reply src/libcprover-cpp/api.cpp:173]: Releasing the file resource between the two API calls has potential concurrency implications. Other system processes have the opportunity to perform various operations in between the call to `is_goto_binary` and `read_goto_binary` which could make the second call fail. I admit this isn't hugely likely, but the possibility of these kinds of race conditions seems worth consideration to me.

Comment (NlightNFotis) [reply src/libcprover-cpp/api.cpp:173]: We know that CBMC is not thread-safe in general, so I admit that this is a possibility, but bar from starting to add locks to various calls like that right now I don't know how to fix it.

I believe that adding locks/making it thread-safe probably falls outside of the scope of this PR. Do you have any suggestions on another way to achieve this, or did you want locks introduced?

Comment (thomasspriggs) [reply src/libcprover-cpp/api.cpp:173]: Thread safety is entirely unrelated issue here. Processes contain threads, the cbmc process has process-level (global) state which would be an issue if we wanted to use multiple threads within the cbmc process. This isn't what this issue is about, as cbmc does not use multiple threads and there is no plan to introduce them to my knowledge. However we should correctly support multi-processing (play-nicely) whether the processes involved are multiple instances of cbmc, or cbmc plus other unrelated user processes. I could go into hypothetical scenarios about how non-related processes could cause issues with the goto-binary in between the API calls but this isn't really the crux of the issue.

I think that having two separate API calls invites incorrect assumptions about how the two API calls relate to each other. Simply removing `is_goto_binary` from the public API would be one way to address this. Potential incorrect assumptions include -

- The assumption that if `is_goto_binary` returns true then the following call to `read_goto_binary` will succeed. This is incorrect even without considering concurrency / race condition scenarios. `is_goto_binary` only checks the header of the file. Therefore `read_goto_binary` could still fail due to error causing data after the header. Therefore the caller of `read_goto_binary` must handle any errors encountered during the API call even if a preceding call to `is_goto_binary` suceeded.
- The assumption that if we are going to call `read_goto_binary` then we must call `is_goto_binary` before hand. As I read the current implementation, `read_goto_binary` includes all the checks which `is_goto_binary` does. Therefore doing a precheck on `is_goto_binary` is unnecessary as it is redundant.

We could keep a publicly facing API call which does a check on whether a particular file is a goto-binary or not. But I don't see what the current advantage of doing so would be?

Comment (NlightNFotis) [reply src/libcprover-cpp/api.cpp:173]: Ah, I think I understand the problem better now. I will paraphrase the concern, to make sure I have understood it correctly - feel free to correct any points that are wrong:

> The issue with `is_goto_binary` and `read_goto_binary` is that the similar naming
> confuses because it implies an implicit sequencing between the two calls that sets
> up certain expectations that can be violated.

If my understanding is correct, would some documentation ameliorate those concerns?

With regards to the API call itself, I agree that it looks a bit out of place, but in general, my opinion is that the API as a whole should provide mechanism, not policy. I can easily picture the API call being part of some future scenarios, something like (pseudocode)

```c
if (is_goto_binary())
   extract_goto_binary_section(relro)
else if (is_elf_binary())
   extract_elf_binary_section(retro)
else
   // don't do anything for source files
```

I don't know how realistic this particular example might be, but scenarios that require some sort of querying of `goto-binary` properties fall within the purview of general API usage (if a bit underdeveloped right now), so I wouldn't want to remove from that capability.

Any thoughts?

File: src/libcprover-cpp/api.h (hunk 20 above)
Comment (thomasspriggs) [reply src/libcprover-cpp/api.h]: This looks to be done.

File: src/libcprover-cpp/api.cpp (hunk 21 above)
Comment (thomasspriggs) [reply src/libcprover-cpp/api.cpp]: This looks to be done.

File: src/goto-bmc/goto_bmc_parse_options.h (hunk 30 above)
Comment (thomasspriggs) [reply src/goto-bmc/goto_bmc_parse_options.h]: This looks to be done.

File: src/goto-bmc/goto_bmc_parse_options.h (hunk 31 above)
Comment (thomasspriggs) [reply src/goto-bmc/goto_bmc_parse_options.h]: This looks to be done.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 29 above)
Comment (thomasspriggs) [reply src/goto-bmc/goto_bmc_parse_options.cpp:72]: Error handling is still missing here. In the case where `goto_bmc_parse_optionst::doit` calls `api.read_goto_binary(filename);` and the goto binary is not successfully read, control flow should not proceed to the `api.run_verifier();` call below this. If we don't have a good solution for this at the moment, then we should probably put a `TODO` here (and raise a follow-up ticket so it isn't forgotten) at the very least.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 32 above)
Comment (thomasspriggs) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: This looks to be resolved.

File: src/libcprover-cpp/api.cpp (hunk 36 above)
Comment (thomasspriggs) [reply src/libcprover-cpp/api.cpp:173]: I think that we should shelve this discussion for potential continuation outside the forum of this PR review, as it was a point for consideration rather than a blocking comment.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 37 above)
Comment (thomasspriggs) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: This appears to be resolved.

File: src/goto-bmc/goto_bmc_parse_options.cpp (hunk 19 above)
Comment (esteffin) [reply src/goto-bmc/goto_bmc_parse_options.cpp]: Fixed

**Review by esteffin — APPROVED** [review]
Change requested has been fulfilled

**Review by thomasspriggs — APPROVED** [review]
The missing error handling has now been added.

File: CODEOWNERS [hunk 39]
```diff
@@ -50,6 +50,7 @@
 # These files change frequently and changes are medium-risk
 
 /src/goto-analyzer/ @martin-cs @peterschrammel
+/src/goto-bmc/ @NlightNFotis @thomasspriggs @esteffin @TGWDB @peterschrammel
```
Comment (thomasspriggs) [inline CODEOWNERS:53]: :+1:

**Review by peterschrammel — APPROVED** [review]
